compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ServiceDesugar.java

```diff
@@ -187,13 +196,26 @@ void rewriteServiceVariable(BLangService service, SymbolEnv env, BLangBlockStmt
         }
     }
 
+    private BType getListenerType(BType type) {
+        if (type.tag == TypeTags.UNION) {
+            for (BType memberType : ((BUnionType) type).getMemberTypes()) {
+                if (types.checkListenerCompatibility(memberType)) {
+                    return memberType;
```

Comment:
Would this work well when the union type has > 1 listener compatible type (say 2), and the user actually assigns a value of the second type to the variable?

Reply:
Updated the test case to test this scenario https://github.com/ballerina-platform/ballerina-lang/pull/27605/files#diff-2d258634ca06bf22db4f17e8bb66906ed1d42a448a1788a8e18403bdc2051583R120

Reply:
```
type LE EmptyListener|Listener;
listener LE lsn = new Listener();
```

In this scenario, this method would return `EmptyListener`, right? Was wondering if that could cause issues, because for example, when we consider the attach method, we'll be retrieving `attach` of `EmptyListener` instead of `Listener` even though the actual value is `Listener`. 

Basic scenarios seem to work as expected though.

Reply:
Not really, since method is invoked on the value as opposed to invoked on type I don't think there will be an issue,
if we can guarantee that we pass the correct types as arguments.
Related discussion: https://github.com/ballerina-platform/ballerina-spec/issues/320